pkg/querier/frontend/frontend_querier_queues.go

```diff
@@ -0,0 +1,227 @@
+// to pass last user index returned by this function as argument. Is there was no previous
+// last user index, use -1.
+func (q *queues) getNextQueueForQuerier(lastUserIndex int, querier string) (chan *request, string, int) {
+	uid := lastUserIndex
+
+	for iters := 0; iters < len(q.users); iters++ {
+		uid = uid + 1
+
+		// Don't use "mod len(q.users)", as that could skip users at the beginning of the list
+		// for example when q.users has shrunk since last call.
+		if uid >= len(q.users) {
+			uid = 0
+		}
+
+		u := q.users[uid]
+		if u == "" {
+			continue
+		}
+
+		q := q.userQueues[u]
+
+		if q.queriers != nil {
+			if _, ok := q.queriers[querier]; !ok {
+				// This querier is not handling the user.
+				continue
+			}
+		}
+
+		return q.ch, u, uid
+	}
+	return nil, "", uid
+}
+
+func (q *queues) addQuerierConnection(querier string) {
+	conns := q.querierConnections[querier]
+
+	q.querierConnections[querier] = conns + 1
+
+	// First connection from this querier.
+	if conns == 0 {
+		q.sortedQueriers = append(q.sortedQueriers, querier)
+		sort.Strings(q.sortedQueriers)
+
+		q.recomputeUserQueriers()
+	}
+}
+
+func (q *queues) removeQuerierConnection(querier string) {
+	conns := q.querierConnections[querier]
+	if conns <= 0 {
+		return
+	}
+
+	conns--
+	if conns > 0 {
+		q.querierConnections[querier] = conns
+	} else {
+		delete(q.querierConnections, querier)
+
+		ix := sort.SearchStrings(q.sortedQueriers, querier)
+		q.sortedQueriers = append(q.sortedQueriers[:ix], q.sortedQueriers[ix+1:]...)
+
+		q.recomputeUserQueriers()
+	}
+}
+
+func (q *queues) recomputeUserQueriers() {
+	scratchpad := make([]string, 0, len(q.sortedQueriers))
+
+	for _, uq := range q.userQueues {
+		uq.queriers = shuffleQueriersForUser(uq.seed, uq.maxQueriers, q.sortedQueriers, scratchpad)
+	}
+}
+
+// Scratchpad is used for shuffling, to avoid new allocations. If nil, new slice is allocated.
+// shuffleQueriersForUser returns nil if queriersToSelect is 0 or there are not enough queriers to select from.
+// In that case *all* queriers should be used.
+func shuffleQueriersForUser(userSeed int64, queriersToSelect int, allSortedQueriers []string, scratchpad []string) map[string]struct{} {
+	if queriersToSelect == 0 || len(allSortedQueriers) <= queriersToSelect {
+		return nil
+	}
+
+	result := make(map[string]struct{}, queriersToSelect)
+	rnd := rand.New(rand.NewSource(userSeed))
+
+	scratchpad = scratchpad[:0]
+	scratchpad = append(scratchpad, allSortedQueriers...)
+
+	last := len(scratchpad) - 1
+	for i := 0; i < queriersToSelect; i++ {
+		r := rnd.Intn(last + 1)
+		result[scratchpad[r]] = struct{}{}
+		// move selected item to the end, it won't be selected anymore.
+		scratchpad[r], scratchpad[last] = scratchpad[last], scratchpad[r]
+		last--
+	}
+
+	return result
+}
+
+func getSeedForUser(user string) int64 {
+	d := md5.New()
+	_, _ = d.Write([]byte(user))
+	buf := d.Sum(nil)
+	return int64(binary.BigEndian.Uint64(buf[:8]) ^ binary.BigEndian.Uint64(buf[8:]))
```

Comment:
Any reason why this logic is different than https://github.com/cortexproject/cortex/pull/3090/files#diff-d8ef8c3e13558a58373d0cbfd02798e0R494-R500 ?

Reply:
Main reason is to avoid dropping half of the result, which seems like a waste.